src/Connections.jl

```diff
@@ -538,25 +610,27 @@ end
 
 function sslupgrade(::Type{IOType}, c::Connection{T},
                     host::AbstractString;
+                    pool::Union{Nothing, Pool}=nothing,
                     require_ssl_verification::Bool=NetworkOptions.verify_host(host, "SSL"),
+                    keepalive::Bool=true,
                     readtimeout::Int=0,
                     kw...)::Connection{IOType} where {T, IOType}
     # initiate the upgrade to SSL
     # if the upgrade fails, an error will be thrown and the original c will be closed
     # in ConnectionRequest
     tls = if readtimeout > 0
-        try_with_timeout(() -> shouldtimeout(c, readtimeout), readtimeout, () -> close(c)) do
-            sslconnection(IOType, c.io, host; require_ssl_verification=require_ssl_verification, kw...)
+        try_with_timeout(readtimeout) do
+            sslconnection(IOType, c.io, host; require_ssl_verification=require_ssl_verification, keepalive=keepalive, kw...)
         end
     else
-        sslconnection(IOType, c.io, host; require_ssl_verification=require_ssl_verification, kw...)
+        sslconnection(IOType, c.io, host; require_ssl_verification=require_ssl_verification, keepalive=keepalive, kw...)
     end
     # success, now we turn it into a new Connection
-    conn = Connection(host, "", 0, require_ssl_verification, tls)
-    # release the "old" one, but don't allow reuse since we're hijacking the socket
-    release(getpool(T), connectionkey(c), c; return_for_reuse=false)
+    conn = Connection(host, "", 0, require_ssl_verification, keepalive, tls)
+    # release the "old" one, but don't return the connection since we're hijacking the socket
+    release(getpool(pool, T), connectionkey(c), nothing)
     # and return the new one
-    return acquire(getpool(IOType), connectionkey(conn), conn)
+    return acquire(() -> conn, getpool(pool, IOType), connectionkey(conn); forcenew=true)
```

Comment:
(timeout of 0 means no timeout; i.e. we'll just wait for as long as possible for the connection to happen)

Reply:
why is it needed here specifically?

Reply:
Oh got it. So in this case, we're "hijacking" the socket, so that's why we return the permit back to the pool, but don't return the connection, but then we still want to abide by the connection pool limits, so we check out a new connection. BUT, we already have the connection, because we created it manually from the socket we hijacked, so we really want to `forcenew` and just return the connection we created manually. The reason we need to return the permit and acquire again anyway is because we're using the "tunneled" host:port as the connection parameter now instead of the original proxy address.